File: src/components/charts/palette.ts
// ── 图表系列色板（spec §A3 定案）──
// 主线 = var(--data-primary)（token `data-primary`，蓝紫 dark #5E6AD2 / light #4E59C4）。
// **不是 var(--primary)**：那是界面填充色，2026-09-23 之后深色下是近黑，
// 画成折线等于看不见。数据色与界面强调色从此是两档，别再合并。
// 辅线灰阶：var(--chart-2..5) 循环。globals.css 的 --chart-N 定义保持纯灰阶不动，
// 色板语义只在这里表达。
import { clamp } from "@/utils/formatters";

export const PRIMARY_COLOR = "var(--data-primary)" as const;

const AUX_COLORS = [
  "var(--chart-2)",
  "var(--chart-3)",
  "var(--chart-4)",
  "var(--chart-5)",
] as const;

/**
 * 第 index 个系列的颜色：0（及负数）→ 主系列琥珀，1 起循环灰阶 --chart-2..5。
 * 例：seriesColor(0)="var(--data-primary)"，seriesColor(1)="var(--chart-2)"，
 * seriesColor(5)= 回绕 "var(--chart-2)"。
 */
export function seriesColor(index: number): string {
  if (index <= 0) return PRIMARY_COLOR;
  return AUX_COLORS[(index - 1) % AUX_COLORS.length];
}

/** 前 count 个系列的颜色数组（ChartConfig / Cell fill 批量取色用）。 */
export function seriesColors(count: number): string[] {
  return Array.from({ length: count }, (_, i) => seriesColor(i));
}

// ── 热力图色带：琥珀 alpha 阶梯（原型验证值 .06 → .92，spec §A3）──
const HEAT_RGB = "232, 197, 71"; // 琥珀 #e8c547
export const HEAT_MIN_ALPHA = 0.06;
export const HEAT_MAX_ALPHA = 0.92;

/**
 * 热力格子颜色：t ∈ [0,1]（格子值 / 色带最大值）→ 琥珀 rgba alpha 阶梯。
 * t 越大越热（越不透明越亮）；区间外 clamp 到两端。
 */
export function heatColor(t: number): string {
  const alpha = HEAT_MIN_ALPHA + (HEAT_MAX_ALPHA - HEAT_MIN_ALPHA) * clamp(t, 0, 1);
  return `rgba(${HEAT_RGB}, ${alpha.toFixed(3)})`;
}

// ── ChartConfig 默认色注入（spec §A3：色板语义在 ChartConfig 层表达）──
import type { ChartConfig } from "@/components/ui/chart";

/**
 * 按 config 键序补默认系列色：第 i 个键未显式给 color（且无 theme 双色定义）时
 * 补 seriesColor(i)（首位琥珀，其余灰阶）。组件统一走这里，再交 ChartStyle 落
 * `--color-<key>`，系列引用 `var(--color-<key>)`。
 */
export function withDefaultColors(config: ChartConfig): ChartConfig {
  return Object.fromEntries(
    Object.entries(config).map(([key, item], i) => [
      key,
      item.color ?? item.theme ? item : { ...item, color: seriesColor(i) },
    ]),
  );
}

File: flutter/lib/src/charts/palette.dart
/// 图表色板（对应 React 版 `src/components/charts/palette.ts`）。
///
/// React 版把主色写成 `var(--primary)` 交给 CSS 解析；Flutter 没有 CSS 变量，
/// 主色从主题取（`AidogTheme.of(context).c`）。辅线灰阶与热力琥珀在 React 侧
/// 是明暗同值的固定色（globals.css / palette.ts），这里直接写死常量，不跟主题。
library;

import 'package:flutter/widgets.dart';

import '../shell/theme.dart';

/// 辅线灰阶四档 = React 的 `--chart-2..5`（globals.css:106-109,811-814），
/// 原值 oklch(0.556/0.439/0.371/0.269 0 0)，明暗两模式同值。
/// oklch→sRGB 换算（2026-09-25 批次一）：零色度 → a=b=0 → 线性值恰为 L³，
/// gamma 编码 1.055·c^(1/2.4)−0.055 后 ×255 四舍五入：
///   0.556³=0.17188 → 0.4515 → 115 → #737373
///   0.439³=0.08460 → 0.3220 →  82 → #525252
///   0.371³=0.05106 → 0.2505 →  64 → #404040
///   0.269³=0.01947 → 0.1493 →  38 → #262626
/// （恰为 Tailwind neutral-500..800，互为佐证。）
const List<Color> kAuxColors = [
  Color(0xFF737373),
  Color(0xFF525252),
  Color(0xFF404040),
  Color(0xFF262626),
];

/// 热力色带端点：与 React 版 `HEAT_MIN_ALPHA` / `HEAT_MAX_ALPHA` 一字不差。
const double kHeatMinAlpha = 0.06;
const double kHeatMaxAlpha = 0.92;

double _clamp01(double t) => t < 0 ? 0 : (t > 1 ? 1 : t);

/// 一套模式下的系列取色。页面不自己算色，一律 `ChartPalette.of(context)`。
@immutable
class ChartPalette {
  const ChartPalette(this.c);

  factory ChartPalette.of(BuildContext context) =>
      ChartPalette(AidogTheme.of(context).c);

  final AidogColors c;

  /// 主系列色（React 版 `PRIMARY_COLOR` = `var(--primary)`）。
  ///
  /// 读 `dataPrimary` 而不是 `accent`：两者曾经同值，直到深色强调色改成近黑
  /// （`#101012`，用户 2026-09-23 定）—— 近黑折线画在 `#111214` 的卡片上是
  /// 1.01:1，**数据本身看不见**。图表是唯一不跟随强调色单色化的地方：多条线
  /// 必须靠色相分开，灰阶只够分 2~3 条（用户同日拍板）。
  Color get primary => c.dataPrimary;

  /// 第 index 个系列的颜色：0（及负数）→ 主色，1 起循环四级灰阶。
  /// 例：`series(0)` = accent，`series(1)` = fg2，`series(5)` 回绕 = `series(1)`。
  Color series(int index) {
    if (index <= 0) return primary;
    return kAuxColors[(index - 1) % kAuxColors.length];
  }

  /// 前 count 个系列的颜色（批量取色）。
  List<Color> seriesColors(int count) =>
      List.generate(count, series, growable: false);

  /// 热力色带色源：固定琥珀 `#e8c547`（React `palette.ts:33` 的 HEAT_RGB），不跟主题。
  static const Color kHeatAmber = Color(0xFFE8C547);

  /// 热力格子颜色：t ∈ [0,1] → 琥珀的 alpha 阶梯，区间外 clamp 到两端。
  /// alpha 公式与 React 版逐字对应（min + (max-min) × clamp(t)）。
  Color heat(double t) => kHeatAmber.withValues(alpha: heatAlpha(t));

  /// 热力格子的 alpha（暴露出来是因为测试按 alpha 断言，与 React 版同一套期望值）。
  ///
  /// 取 3 位小数：React 版落到 CSS 时写的是 `alpha.toFixed(3)`，那一步同时也吃掉了
  /// 浮点累加噪声（0.92 会算成 0.9200000000000002）。这里复刻同一个精度。
  static double heatAlpha(double t) => double.parse(
    (kHeatMinAlpha + (kHeatMaxAlpha - kHeatMinAlpha) * _clamp01(t))
        .toStringAsFixed(3),
  );
}
